Poll Nissan servers for battery updates

bwduncan opened the pull request:
## Proposed change
Fix the battery status update so that it polls until a response has been received from the car.

This restores the previous behaviour of checking the returned timestamp, but also issues a new request if the Nissan servers return success but without changing the timestamp. This reliably updates from my car once an hour, every time.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! # Unexpected factorial!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

bwduncan committed:
Poll Nissan servers for battery updates

Restore the previous behaviour of checking the "OperationDateAndTime" value in the returned battery status, issuing a new battery check request if it hasn't changed. Also add the sleeps as discovered by @irekzielinski

bwduncan committed:
Remove unnecessary 'else' for pylint

bwduncan committed:
Check for start_date is None

Nissan may not return a start date, which is fine, but in that case we just have to return the data we have, instead of trying another 9 times and then returning it anyway.

hencoappel commented on homeassistant/components/nissan_leaf/__init__.py:
```diff
@@ -367,7 +372,20 @@ async def async_get_battery(self):
                     )
-                    return server_info
+                    if (
+                        start_date is not None
```
Could you simplify this to:
`if start_date and start_date != self._extract_start_date(server_info):`

bwduncan replied:
Yes I think you're right. I will make this change and test for a few hours.

Thanks!

bwduncan committed:
Simplify start date check

`start_date is not None` can be simplified to just `start_date` thanks to @hencoappel

Move static method `_extract_start_date` out of `LeafStore` since it doesn't really belong there (Maybe it should move to pycarwings2?) and to keep black happy. Makes the code look much cleaner to me.

bwduncan committed:
Check get_latest_battery_status return

A few days of testing reveals that this method can return None. In an ideal world this function would just raise an exception, but here we are...

bwduncan committed:
More failure modes

pycarwings2 can raise an exception if Nissan returns an empty list where battery records should be. My car stopped talking to me this week and this got the component completely stuck. Guard against this.

I got the logic wrong. If we failed to get an initial "start_date", we can't know if the data has been refreshed, we just have to trust the Nissan servers and return the first data they give us. (OR, if we did get a start_date, wait until it has changed)

bwduncan commented:
Just leaving myself a little note here that pycarwings2 doesn't pass a timeout to requests so if there is a network failure it appears that the whole lot just hangs forever (23 hours so far...). I'm not entirely sure what the point of the `update_in_progress` flag is, it seems to just make the whole thing stop if any uncaught exception is raised, but maybe there is a good reason for it. I'm going to test giving requests a timeout and see if that fixes the problem, and if it does I'll make another pull request.

bwduncan commented:
> There hasn't been any activity on this pull request recently. This pull request has been automatically marked as stale

This PR is not stale it is waiting for review, ideally by @filcole the author of the component.

bdraco commented on homeassistant/components/nissan_leaf/__init__.py:
```diff
@@ -326,20 +334,25 @@ async def async_refresh_data(self, now):
             _LOGGER.debug("Requesting battery update, %s", self.leaf.vin)
+            start_date = None
+            try:
+                start_server_info = await self.hass.async_add_executor_job(
```
Can we store this between updates so we don't have to fetch it each time?

bwduncan replied:
I don't think so. In my testing, if this call isn't made (with the following 1 second sleep) then the API just never returns anything, we retry ten times and then give up. I believe doing this closely mimics the behaviour of the official app.

bdraco commented on homeassistant/components/nissan_leaf/__init__.py:
```diff
@@ -326,20 +334,25 @@ async def async_refresh_data(self, now):
+                else:
+                    start_date = _extract_start_date(start_server_info)
+            await asyncio.sleep(1)  # Critical sleep
             request = await self.hass.async_add_executor_job(self.leaf.request_update)
```
This isn't polling too frequently so if we stored the last value, we could request the update and then schedule a check 60 seconds later to get the new data and check to see if its changed

bwduncan replied:
Can you explain a bit more what you mean? Is not polling too frequently a good thing?

I am open to scheduling checks and waiting for them in other methods, but my intention here is to make the code work without diverging too much from @filcole 's creation.

bdraco reviewed (changes requested):
Please see comments above.  Everything here looks like its very async.  We should fire off the request for update and check later instead of hitting the api over and over.

bwduncan commented:
This isn't stale, I just haven't had time to look at the review comments this week. I could do with some help understanding them, and I would welcome input from the component author.

MartinHjelmare commented:
There's been no activity here for more than a month. Are you planning to continue now? Otherwise there's no point in keeping this open. If you don't have time now, please open a new PR when you're ready to continue.

If you have questions on the review please ask specific questions.

bwduncan commented:
Hi Martin and Nick,

Thanks so much for giving this PR some attention. I felt like I was talking into the void!

I am prepared to continue now (or probably, after the easter weekend, it depends on the weather!). I am happy to test the code on my HA instance with my car. It would be nice to get others input (especially the author of the component) but I believe my changes make the code actually work in my situation.

I am prepared to make the changes Nick alluded to about leveraging the async nature of the requests, but I am fairly new to async so I would appreciate some help.

Thanks!
Bruce

bwduncan committed:
Check server_info, catch TypeError

server_info can be None, so make sure that it is valid before passing it to _extract_start_date.

If you leave the car for a couple of weeks, the API behaves even more strangely. Later calls to get_latest_battery_status will also fail, and there's nothing we can do about that, so just add TypeError to the list of exceptions which will log an error and return None

bwduncan commented:
Please can someone merge this? I didn't write the original component and I feel like it would be risky to make big changes to it. The changes I am proposing improve the reliability of the component without a massive rewrite. I'll be happy to contribute to further cleanups but I think it's important to get the component working reliably as a starting point.

bdraco commented:
> Please can someone merge this? I didn't write the original component and I feel like it would be risky to make big changes to it. The changes I am proposing improve the reliability of the component without a massive rewrite. I'll be happy to contribute to further cleanups but I think it's important to get the component working reliably as a starting point.

The current implementation is solving the issue by increasing the load on Nissan's servers which is probably something they won't be happy about.  We want to minimize the load on vendors servers as much as possible as it reflects poorly on Home Assistant when integrations poll frequently and sometimes results in vendors blocking APIs or banning users.  

I've put together an rough example of a potential solution that:

1. Requests the update
2. Schedules checking 60 seconds later to ensure enough time has passed for the update to complete

https://github.com/bdraco/home-assistant/commit/82c64d302c2fed629b34adf8185063ae89502dc8

bdraco commented:
MAX_RESPONSE_ATTEMPTS is currently set at 10.  That's currently ~30 additional api calls that can happen per update.  Would it be reasonable to set this to something lower and accept the old data rather than generating the additional calls?

It looks like the integration allows setting the update interval as low as 2 minutes. That seems like its going to be quite unfriendly to Nissan's infrastructure if the user chooses a value that low.

Usage of this integration isn't that high though per https://analytics.home-assistant.io/#integrations so the overhead likely isn't that high today, but could become an issue with wider adoption. Generally once a vendor decides to start blocking users, and already invested the time in building the infra to rate limit or block non-official integrations, we see integrations disappear so we want to be as respectful as possible to avoid this type of situation.

bwduncan commented:
That's an interesting point. I had a look in my logs. Sadly I rebooted my raspberry pi yesterday and so I only have one day of home assistant logs. However, in 18 updates, almost all of them required 0 or 1 retries. There were two which retried 10 times and then gave up. On that basis I think you're right that it probably would be safe to lower `MAX_RESPONSE_ATTEMPTS` for me, but bear in mind this was only a snapshot in time for one user with one car on one day. I'd also like to point out that this retry logic wasn't created by me, and the load it generates is low on average.

I think most users probably don't change the defaults. There are also warnings in the docs and code telling people not to lower the refresh rate too much. I'm not sure what you would gain by lowering the interval: Most of the time cars are just parked. If the car is charging we automatically decrease the polling interval, which makes the graphs look pretty and helps to decide when to stop charging.

That's fascinating analytics, thanks for pointing it out! Nissan do occasionally release a new version of their app which breaks backward compatibility, but I don't think they are really bothered about third party users (there are other mobile apps which replicate the functionality of the official app but without the glitches. Have a look at the reviews! https://play.google.com/store/apps/details?id=com.digitas.android.nissan.carwings&hl=en&gl=US). As far as new users are concerned, I think we have probably reached a plateau. I believe the app only works with car models up to 2019 (model ZE0 generation 1) so we'll maybe get a few new users like myself who bought second-hand cars but that will probably be balanced by cars being taken off the road.

Once again, thank you for engaging with me on this PR.

bdraco commented:
> That's an interesting point. I had a look in my logs. Sadly I rebooted my raspberry pi yesterday and so I only have one day of home assistant logs. However, in 18 updates, almost all of them required 0 or 1 retries. There were two which retried 10 times and then gave up. On that basis I think you're right that it probably would be safe to lower `MAX_RESPONSE_ATTEMPTS` for me, but bear in mind this was only a snapshot in time for one user with one car on one day. I'd also like to point out that this retry logic wasn't created by me, and the load it generates is low on average.
> 

I'd be comfortable merging this if the value was adjusted down to 2 or 3.  If there is a change to the api later, and suddenly its now taking 8-10 attempts we won't be able to get users who currently have the integration deployed to update immediately, but at least we limit the fallout.

bwduncan commented:
OK cool. I would be happy enough with 3. :+1: 

Ideally we should detect when the Nissan servers are being uncooperative and back off. They seem to stop trying after a while. For example, because my car hasn't moved for a week, the updates never complete. Retrying 10 times every hour in this case is pointless. However I don't think the complexity of a backoff strategy is appropriate for this PR.

bwduncan committed:
Limit MAX_RESPONSE_ATTEMPTS to 3

If the Nissan servers aren't responding, it seems unlikely that more than 3 retries will be helpful. If they are really stuck, reducing the limit reduces the load on the servers by ~3. Sometimes, the servers need a delay of about a minute to get the data, so at least 2 retries are often necessary.

bwduncan commented:
Hey @bdraco sorry for the delay, life got busy. I'm happy with the state of the PR now, if you are.

frenck commented:
Not stale, awaiting our review.

bdraco reviewed (approved):
Sorry I missed your response.  LGTM
